src/engine/engine.cpp

```diff
@@ -393,6 +394,57 @@ Engine::pgzStructure_t Engine::getPartGroupZoneStructure(int partFilter) const
     return res;
 }
 
+void Engine::loadSampleIntoZone(const fs::path &p, int16_t partID, int16_t groupID, int16_t zoneID,
+                                int sampleID, int16_t rootKey, KeyboardRange krange,
+                                VelocityRange vrange)
+{
+    assert(messageController->threadingChecker.isSerialThread());
+    assert(sampleID < maxSamplesPerZone);
+
+    // TODO: Deal with compound types more comprehensively
+    // If you add a type here add it to Browser::isLoadableFile also
+    if (extensionMatches(p, ".sf2") or extensionMatches(p, ".sfz"))
+    {
+        assert(false);
+        return;
+    }
+
+    auto sz = getSelectionManager()->currentLeadZone(*this);
+
+    if (!sz.has_value())
+    {
+        messageController->reportErrorToClient("Unable to load Sample",
+                                               "There is no currentLeadZone");
+        return;
+    }
+
+    assert((*sz).part == partID);
+    assert((*sz).group == groupID);
+    assert((*sz).zone == zoneID);
+
+    // OK so what we want to do now is
+    // 1. Load this sample on this thread
+    auto sid = sampleManager->loadSampleByPath(p);
+
+    if (!sid.has_value())
+    {
+        messageController->reportErrorToClient(
+            "Unable to load Sample",
+            "Sample load failed:\n\n" + p.u8string() + "\n\n" +
+                "It is either an unsupported format or invalid file. "
+                "More information may be available in the log file (menu/log)");
+        return;
+    }
+
+    auto &part = getPatch()->getPart(partID);
```

Comment:
Hi! So I reviewed everything before I take off for the day to my meetings, and it all looks good (although theres that MSVC error which CI tickled). The only bit I worry about is these lines below (440-444) which I think need to be on the audio thread.

so I would change this to a 

```
cont.scheduleAudioThreadCallback([groupId, zoneID, samp = *sid](engine &e)
{
auto &zone = eng.getPatch()->getPart(partId)->getGroup(gId)->getZone(zoneId;
zome->sampledata = etc
zone->attachToSample
},
[p,g,z](auto e) { selectionManager->selectAction() }
);
```

this way we guarantee the loading and UI happens off the audio thread but the structural mutation happens on the thread, which should avoid conflicts if you do this modification when voices are playing.

We may *also* want to consider killing voices when we do this action. I added an API to do that last week `Zone::terminateAllVoices`

Otherwise looks great to me!

Reply:
Shall I use the `scheduleAudioThreadCallbackUnderStructureLock` method as it is used in `loadSampleIntoSelectedPartAndGroup`?

Reply:
That’s the more conservative choice. I don’t think this attachment modifies the structure but I would have to read the code. For now it’s safer to do so though yes